data_model/primitives/src/conststr.rs

```diff
@@ -0,0 +1,545 @@
+//! Const-string related implementation and structs.
+#[cfg(not(feature = "std"))]
+use alloc::{boxed::Box, str::from_boxed_utf8_unchecked, string::String, vec::Vec};
+use core::{
+    cmp::{Eq, Ord, Ordering, PartialEq, PartialOrd},
+    convert::{AsRef, From, TryFrom},
+    fmt::{Debug, Display, Formatter, Result as FmtResult},
+    hash::{Hash, Hasher},
+    mem::{size_of, ManuallyDrop},
+    ops::Deref,
+};
+#[cfg(feature = "std")]
+use std::str::from_boxed_utf8_unchecked;
+
+use iroha_schema::{IntoSchema, MetaMap};
+use parity_scale_codec::{Encode, Output};
+use serde::{Serialize, Serializer};
+
+const MAX_INLINED_STRING_LEN: usize = 2 * size_of::<usize>() - 1;
+
+/// Immutable inlinable string.
+/// Strings shorter than 15/7/3 bytes (depending on architecture 64/32/16 bit pointer width) are inlined.
+#[derive(Clone)]
+pub struct ConstString(ConstStringData);
+
+impl ConstString {
+    /// Returns the length of this [`ConstString`], in bytes.
+    #[inline]
+    pub fn len(&self) -> usize {
+        self.0.len()
+    }
+
+    /// Returns `true` if [`ConstString`] is empty.
+    #[inline]
+    pub fn is_empty(&self) -> bool {
+        self.0.len() == 0
+    }
+
+    /// Returns `true` if [`ConstString`] data is inlined.
+    #[inline]
+    pub const fn is_inlined(&self) -> bool {
+        self.0.is_inlined()
+    }
+
+    /// Construct empty [`ConstString`].
+    #[inline]
+    pub const fn new() -> Self {
```

Comment:
check out `derive_more::Constructor`

Reply:
`derive_more::Constructor` implements `new` without `const` and also uses struct fields as constructor arguments which we don't want in this case.